Fixes phase follow

juleluj opened the pull request.

juleluj committed:
fixes phase follow

miili commented on lightguide/blast.py:
```diff
@@ -330,6 +331,8 @@ def follow_phase(
         threshold: float = 5e-1,
         max_shift: int = 20,
+        stack: bool = False,
+        no_of_stacks: int = 5,
```
Merge these two parameters.

miili commented on lightguide/blast.py:
```diff
@@ -409,14 +427,21 @@ def correlate(data: np.ndarray, direction: Literal[1, -1] = 1) -> None:
                 template = trace[
                     phase_idx - window_size[0] : phase_idx + window_size[1] + 1
                 ].copy()
+
+                # stacking
+                if len(template_deque) <= no_of_stacks:
+                    template_deque.append(template)
+                if len(template_deque) == no_of_stacks + 1:
+                    template_deque.popleft()
+                template_stack = np.sum(template_deque, axis=0) / len(template_deque)
+
```
Move stacking to `prepare_template`

miili commented on lightguide/blast.py:
```diff
@@ -375,12 +384,21 @@ def follow_phase(
         def correlate(data: np.ndarray, direction: Literal[1, -1] = 1) -> None:
             template = root_template.copy()
-            index = root_idx
+            template_deque = deque([np.array(template)])
```
Use `maxlen` from `deque` and skip the `pop`

miili commented on lightguide/blast.py:
```diff
@@ -395,7 +413,7 @@ def correlate(data: np.ndarray, direction: Literal[1, -1] = 1) -> None:
                 phase_time = self._sample_to_time(int(phase_idx))
 
-                if phase_correlation < threshold:
+                if phase_correlation > threshold:
```
revert

juleluj committed:
fixes follow phase

juleluj committed:
modified & added example

miili commented on lightguide/blast.py:
```diff
@@ -330,6 +333,7 @@ def follow_phase(
         window_size: int | tuple[int, int] = 50,
         threshold: float = 5e-1,
         max_shift: int = 20,
+        no_of_stacks: int = 1,
```
Rename to `template_stacks`?

miili commented on lightguide/blast.py:
```diff
@@ -373,14 +382,18 @@ def follow_phase(
         pick_channels, pick_times, pick_correlations = [], [], []
 
         def prepare_template(data: np.ndarray) -> np.ndarray:
+            data = np.sum(data, axis=0) / len(data)
```
use `np.mean`?

miili commented on lightguide/blast.py:
```diff
@@ -373,14 +382,18 @@ def follow_phase(
         pick_channels, pick_times, pick_correlations = [], [], []
 
         def prepare_template(data: np.ndarray) -> np.ndarray:
```
signature: `data: Deque[np.ndarray]`?

juleluj committed:
added function to trim from a Picks-object to blast
added spectrogram function to calculate spectrograms
+ example notebook for it
updated client.py to download different data set
added tracefeatures (maybe integrate differently)
updated examples

miili commented on lightguide/blast.py:
```diff
@@ -340,7 +346,7 @@ def follow_phase(
         3. Evaluate maximum x-correlation in allowed window (max_shift).
         4. Update template trace and go to 2.
-
+            4a. if no_of_stacks >1: stack templates for correlation to stabilize
```
rename

miili commented on lightguide/blast.py:
```diff
@@ -785,7 +832,7 @@ def __len__(self) -> int:
+    # why some functions two times??
     trim_time = shared_function(Blast.trim_time)
     trim_channels = shared_function(Blast.trim_channels)
 
```
Please delete

miili commented on lightguide/client.py:
```diff
@@ -31,3 +34,24 @@ def vsp_shot(cls) -> Blast:
+        return Blast.from_miniseed(file, channel_spacing=1.0)
+
+    @classmethod
+    def markerfile(cls) -> Blast:
```
wrong annotation.

miili commented on lightguide/models/picks.py:
```diff
@@ -0,0 +1,51 @@
+
+
+class Picks(BaseModel):
+    channel: list[int] = [-1]
```
remove defaults?

miili commented on lightguide/models/picks.py:
```diff
@@ -0,0 +1,51 @@
+    correlation: list[float] = []
+    kind: list[int] = []
+
+    def save_picks(self, filename):
```
use `Path` for filename

miili commented on lightguide/models/picks.py:
```diff
@@ -0,0 +1,51 @@
+            markers.append(m)
+        marker.save_markers(markers=markers, filename=filename)
+
+    def get_pyrocko_picks(self, filename):
```
use

```py
@classmethod
def get_pyrocko_picks(cls, filename: Path):
    ...

# Usage:
Picks.from_pyrocko_picks(file)
```

miili commented on lightguide/models/picks.py:
```diff
@@ -0,0 +1,51 @@
+            filename (str): filename to read
+        """
+        markers = marker.load_markers(filename)
+        channels = [m.nslc_ids[0][1] for m in markers]
```
Use `%05d`...

juleluj committed:
modifications

juleluj committed:
modified

juleluj committed:
fixed proposed changes

miili commented on lightguide/models/picks.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class Picks(BaseModel):
+    channel: list[int] = []
+    time: list[datetime] = []
```
Remove defaults

miili commented on lightguide/models/picks.py:
```diff
@@ -0,0 +1,54 @@
+    correlation: list[float] = []
+    kind: list[int] = []
+
+    def save_picks(self, filename: PathStr):
```
Add missing annotations.

miili commented on lightguide/models/picks.py:
```diff
@@ -0,0 +1,54 @@
+
+        return Picks(channel=channels, time=times, kind=kinds)
+
+    # def plot ()
```
remove

miili commented on lightguide/blast.py:
```diff
@@ -409,15 +423,19 @@ def correlate(data: np.ndarray, direction: Literal[1, -1] = 1) -> None:
-        pick_channels = np.array(pick_channels) + self.start_channel
-        pick_correlations = np.array(pick_correlations)
+        pick_channels = (np.array(pick_channels) + self.start_channel).tolist()
 
-        return pick_channels, pick_times, pick_correlations
+        return Picks(
+            channel=pick_channels, time=pick_times, correlation=pick_correlations
+        )
```
Rather to the `.tolist` when you create `Pick`

juleluj committed:
added average traces

juleluj committed:
added compatibility/conversion with obspy & snuffler
added trace management (delete,reduce traces) via 'channel_list'

miili commented on lightguide/blast.py:
```diff
@@ -129,18 +135,67 @@ def duration(self) -> float:
+        idx = self.get_channel_index(channel, strict=True)
+        if idx == None:
+            print(f"#{channel} not in list.")
+            return self
```
Rather `raise ValueError(f"channel {channel} is not in range")`

miili commented on lightguide/blast.py:
```diff
@@ -129,18 +135,67 @@ def duration(self) -> float:
+        """
+        return self.channel_list[channel_index]
+
+    def get_channel_index(self, channel: int, strict=False) -> int:
```
Be always strict. Or is there advantage in raising an exception only sometimes?

miili reviewed (changes requested):
Great, thanks Juliane!

juleluj committed:
fixed problems in channel handling
added co-author

juleluj committed:
-blast.py
    improved phase follower:
        initial pick is lightguide picks object
        correlation window of trace has twice the length of template window
    added ability to plot picks of blast directly
-picks.py
    add/join multiple picks objects
    sort picks object by attriute
-3-event-analysis.ipynb
    updated examples as phase follower uses different pick format
    added example of blast/trace visualization

juleluj committed:
added function to get times of pick at channel
